hail/python/hailtop/utils/utils.py

```diff
@@ -362,41 +405,43 @@ async def run_and_cleanup():
             try:
                 async with self._subsema:
                     await f(*args, **kwargs)
+            except asyncio.CancelledError:
+                pass
             except:
-                if not self._exception:
+                if self._exception is None:
                     _, exc, _ = sys.exc_info()
                     self._exception = exc
                     await self._shutdown()
+                else:
+                    log.info('discarding exception', exc_info=True)
```

Comment:
Do we want this message in user logs? Or will that be too confusing?

Reply:
I'm not sure what to do here.  Once we introduce concurrency, multiple exceptions can happen simultaneously, and the question is how to resolve that.  I see a few options:

1. (This is the first path I took.)  Raise the first exception and log (using `info`) any secondary exceptions.  I felt like this was OK because the user always got the first/primary exception.  (You might argue about which exception I'm discarding.  For example, in aexit, I discard the the task manager exception in favor of an earlier task exception, and maybe we should switch that?)

2. Log the exceptions more forcefully (with `error`, say).  Problem is, I don't really know if it is an error.  This can also potentially generate a large number of exception messages if there are a large number of tasks running concurrently, and I wanted to avoid that (although maybe we shouldn't).

3. Gather all the exceptions together and raise them with a new exception type, like `MultipleExceptions(exceptions)`.  This has the benefit that it doesn't loose information, except you loose all knowledge about the original exception types so callers can't reasonably use try/except, etc.

Thoughts @danking?  It seems like it would be good to have a standard convention for this case.

Reply:
I'm going to go ahead and approve this because Lindo is waiting on this fix. We should revisit this discussion later though.